Design Document for `eth_subscribe` and `eth_unsubscribe` (#896)
hashgraph/hedera-json-rpc-relay: docs/design/eth-subscribe.md

1. Maintain a list of all live socket connections. This can be done with a `key-value` store where the `key` is a unique identifier (some ws libraries provide such an id, but if needed it can be generated by the relay) of the socket object, and the `value` is the socket object itself.

2. Maintain a list of all active subscriptions and the corresponding socket objects. It should have the following structure:

```javascript
  subscriptions: [
    {
      method: string, // newHeads, logs, etc
      params: any,
      subscribers: [{   // a list of all active subscriptions to the specific `method + params` combination
        socketId: string,
        subscription: string
      }],
      lastUpdated: BlockNumber // The latest block number at which this subscription was updated
    }
  ]

```

3. Create a scheduled job that loops through all saved subscriptions and calls the Mirror node Rest API and sends the response to all `subscribers`. It should be possible to configure the time interval. A throttling mechanism can be implemetented that limits the consecutive calls to the Mirror node, and at every interval only X subscriptions with the oldest `lastUpdated` should be polled.

> ## Error Codes

Reviewer: There should be an error case also for
- max subscription reached
- duplicate subscription id
- non matching subscription params e.g if no addresses are matched to any existing contracts
- subscription aborted (maybe when the relay is forced shut down - can we do this?)
- ... (any more)

| Error Codes |                                                       Error message                                                        |                                    Solution                                     |
| :---------: | :------------------------------------------------------------------------------------------------------------------------: | :-----------------------------------------------------------------------------: |
|    32600    | "The maximum batch size that can be sent over a websocket connection is 10. Please decrease the batch size and try again.” | Occurs when user attempts to send high-volume JSON-RPC traffic over Websockets. |

## Limits:

> 1. WebSockets in general are resource intensive. We could implement a throttling mechanism. Every live socket connection would also result in additional requests to the Mirror Node every n seconds. This means that we need to introduce a rate limiting feature, like 5 concurrent connections by IP.

Reviewer: 100%, let's provide a suggestion for a max number of socket connections and maybe the max subscription time to ensure resources aren't locked up for too long.
The last suggestion might be problematic but let's explore it

## Tasks (in suggested order):

#### Milestone 1 (done)

1. Finalize design document

#### Milestone 2

1. Implement library for websocket support in the server.
2. Extend current koa server logic.
3. Introduce `eth_subscribe` and `eth_unsubscribe` to both server and relay part.

> 4. Add logic to save all subsciption IDs and it's corresponding filter to a cache.

Reviewer: Add a step for filter param support.
1st maybe single address, then topics, then multiple contracts etc.
Just a suggestion but I'd like us to take a staged approach here.

5. Add scheduled job function for looping through all saved filters and returns result to it's corresponding socket with the respective subscription id.
6. Extend current ratelimiter class to support websocket limitation.
7. Add logic to delete subsciption upon connection break or request.

> #### Milestone 3 (use Mirror-Node GraphQL API)

Reviewer: Note that the Mirror Node has not yet design this feature at the time of writing and there is no current schedule for when the relay would adopt this

1. Remove scheduled job, as this is going to be handled by the mirror-node graphql module.

> 2. Forward all request for every coresponsing subscription id to the mirror node and the response back to the requester.

Reviewer: ```suggestion
2. Forward all requests for every corresponding subscription id to the mirror node and the return the response back to the requester.
```

# Subscription design

## Purpose

> The JSON-RPC Relay currently doesn't support subscription using websocket as is common in many other relays and is expected by many web3 tool, which makes polling for event logs, new blocks and transactions very hard for developers.

Reviewer: nit: maybe i missed this in my previous suggestion
```suggestion
The JSON-RPC Relay currently doesn't support subscription using websocket as is common in many other relays and is expected by many web3 tools, which makes polling for event logs, new blocks and transactions very hard for developers.
```

## Goals

1. Implement websocket subscription in the relay
2. Introduce two new methods - `eth_subscribe` and `eth_unsubscribe`
3. Support subscription for event logs

## Non-Goals

1. Add long-term memory to the relay

> 2. Reduce the efficiency of the relay or mirror node due to high resource usage

Reviewer: ```suggestion
2. Reduce the efficiency of the relay or mirror node due to high resource usage
3. Support subscription for new blocks and pending transactions
```

## Architecture

### Subscription Types

|   Subscription Type    |                                  Description                                   |           Support           |
| :--------------------: | :----------------------------------------------------------------------------: | :-------------------------: |
|          logs          |      Emits logs attached to a new block that match certain topic filters.      |       Yes, initially        |
|        newHeads        |               Emits new blocks that are added to the blockchain.               | No, maybe in future release |

> | newPendingTransactions | Emits transaction hashes that are sent to the network and marked as `pending`. | No, maybe in future release |

Reviewer: Probably not right for this suggested reason
```suggestion
| newPendingTransactions | Emits transaction hashes that are sent to the network and marked as `pending`. | No, as Hedera does not have pending transactions on a node. |
```

### Initiating a subscription

#### Request

```javascript
// initiate websocket stream first

> wscat -c wss://testnet.hashio.io/api

Reviewer: Don't hardcode testnet, so people know they can configure it per their needs
```suggestion
wscat -c wss://<env>.hashio.io/api
```

## Non-Functional Requirements

WebSockets in general are resource intensive. We could implement a throttling mechanism. Every live socket connection would also result in additional requests to the Mirror Node every n seconds. This means that we need to introduce a rate limiting feature, like 5 concurrent connections by IP.

1. Implement an IP based limit for maximum concurrent connections.
2. Implement a global limit for maximum concurrent connections.
3. Explore the possibility of adding a TTL limit for every connection.
4. Explore the option to add a limit for the total subscribed time per IP ( `sum(connection * subscribed_time)` ). There could theoretically be a way to charge users to increase this limit, for example through via a smart contract. Alternatively users could be encouraged to run their own instances of the relay.
5. Users should be required to renew their subscription in the case of an error. If for some reason the relay is restarted it should not automatically try to restore all previous connections and subscriptions.

## Open Questions

1. How will we limit connections and subscriptions?
2. Is it reasonable to add a TTL timer to subscriptions?

> 

Reviewer 2: Have we thought through the expectations around subscriptions and performance?  Ethereum's block times are approximately 13 seconds.  Will this architecture support the polling of the mirror node every 13 seconds with out triggering rate limits and throttling?  I am asking because some of this transaction related event information can be extremely time sensitive when maintaining asset prices on various protocols, e.g. arbitrage between liquidity pools, oracles adjusting prices based on external events, etc.

Reviewer 3: I'm not sure what the exact rate limits are for the mirror node, but I did a quick test and it can easily handle at least 100 requests per second.

Reviewer 2: Okay, that's good news.

## Non-Functional Requirements

WebSockets in general are resource intensive. We could implement a throttling mechanism. Every live socket connection would also result in additional requests to the Mirror Node every n seconds. This means that we need to introduce a rate limiting feature, like 5 concurrent connections by IP.

1. Implement an IP based limit for maximum concurrent connections.
2. Implement a global limit for maximum concurrent connections.
3. Explore the possibility of adding a TTL limit for every connection.
4. Explore the option to add a limit for the total subscribed time per IP ( `sum(connection * subscribed_time)` ). There could theoretically be a way to charge users to increase this limit, for example through via a smart contract. Alternatively users could be encouraged to run their own instances of the relay.
5. Users should be required to renew their subscription in the case of an error. If for some reason the relay is restarted it should not automatically try to restore all previous connections and subscriptions.

> 

Reviewer 4: Should we also explore the possibility of auto-ending some subscriptions when they are responded, for example:
 1. User subsribes to a transaction by hash, that is on pending state.
 2. Relay receives the transaction once is mined and included in a block and returns the value to the subscriber.
 3. since that tx is final, is certain there wont be more changes, user should unsubscribe, but in case he does not, we automatically do at the time of response.

Reviewer 4: I already answered this from the meeting.

## Open Questions

> 1. How will we limit connections and subscriptions?

Reviewer: Have we answered this or is it still open?
The `Non-Functional Requirements` above should answer some of this right?

|          logs          |      Emits logs attached to a new block that match certain topic filters.      |                       Yes, initially                        |
|        newHeads        |               Emits new blocks that are added to the blockchain.               |                 No, maybe in future release                 |
| newPendingTransactions | Emits transaction hashes that are sent to the network and marked as `pending`. | No, as Hedera does not have pending transactions on a node. |

### Initiating a subscription

#### Request

```javascript
// initiate websocket stream first
wscat -c wss://<env>.hashio.io/api

// then call subscription
{"jsonrpc":"2.0","id": 1, "method": "eth_subscribe", "params": [SUBSCRIPTION_TYPE, PARAMS]}
```

#### Response

```javascript
{"jsonrpc":"2.0","id": 1, "result": SUBSCRIPTION_ID}
```

### Logs

     The `logs` subscription type emits logs that match a specified topic filter and are included in newly added blocks.

#### Parameters

An object with the following fields:

> - address (optional): [`string`] or [`array of strings`] Singular address or array of addresses. Only logs created from one of these addresses will be emitted.

Reviewer: As part of release we can prioritize `string` and have `array of strings` follow in a future release.
These should be separate tickets

#### Request

```javascript
// initiate websocket stream first
wscat -c wss://<env>.hashio.io/api

// then call subscription
{"jsonrpc":"2.0","id": 1, "method": "eth_subscribe", "params": ["logs", {"address": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48", "topics": ["0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef"]}]}
```

#### Response

This response is sent whenever new data is available.

```javascript
{
  "method": "eth_subscription",
  "params": {
    "subscription": SUBSCRIPTION_ID,
    "result": {
      "address": "0x8320fe7702b96808f7bbc0d4a888ed1468216cfd",
      "blockHash": "0x61cdb2a09ab99abf791d474f20c2ea89bf8de2923a2d42bb49944c8c993cbf04",
      "blockNumber": "0x29e87",
      "data": "0x00000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000003",
      "logIndex": "0x0",
      "topics": ["0xd78a0cb8bb633d06981248b816e7bd33c2a35a6089241d099fa519e361cab902"],
      "transactionHash": "0xe044554a0a55067caafd07f8020ab9f2af60bdfe337e395ecd84b4877a3d1ab4",
      "transactionIndex": "0x0"
    }
  }
}
```

> ### newHeads

Reviewer: Drop the `newHeads` and `newPendingTransactions` sections since they aren't part of the design

       "stateRoot": "0xb3346685172db67de536d8765c43c31009d0eb3bd9c501c9be3229203f15f378",
       "timestamp": "0x56ffeff8",
       "transactionsRoot": "0x0167ffa60e3ebc0b080cdb95f7c0087dd6c0e61413140e39d94d3468d7c9689f"
     },
   "subscription": SUBSCRIPTION_ID
   }
 }
```

### newPendingTransactions

     The newPendingTransactions subscription type subscribes to all pending transactions via WebSockets (regardless if you sent them or not), and returns their transaction hashes.

Returns the hash for all transactions that are added to the pending state (regardless if you sent them or not).

#### Request

```javascript
// initiate websocket stream first
wscat -c wss://<env>.hashio.io/api

// then call subscription
{"jsonrpc":"2.0","id": 2, "method": "eth_subscribe", "params": ["newPendingTransactions"]}
```

#### Response

This response is sent whenever new data is available.

```javascript
{"id":1,"result":"0xc3b33aa549fb9a60e95d21862596617c","jsonrpc":"2.0"}

{
    "jsonrpc":"2.0",
    "method":"eth_subscription",
    "params":{
        "subscription": SUBSCRIPTION_ID,
        "result":"0xd6fdc5cc41a9959e922f30cb772a9aef46f4daea279307bc5f7024edc4ccd7fa"
    }
}
```

> ## Mock code

Reviewer: If this mock code was inspired we should credit the inspiration source here.

Reviewer 3: It wasn't inspired, we wrote it when researching how `eth_subscribe` can be implemented.

Reviewer: Great, thanks

      const event = params[0];

      if (event === 'log') {
        const filter = params[1];
        // Cache the `event`, `filter`, `SUBSCRIPTION_ID` and the current block number values

        // This establishes the subscription
        ws.send(
          JSON.stringify({
            id: request.id,
            result: SUBSCRIPTION_ID,
            jsonrpc: '2.0',
          })
        );
      }

      // Handle other types of events: newPendingTransactions, newHeads, syncing
    }

    // The WebSocketProvider emits an event to `eth_chainId` right after establishing a connection
    else if (method === 'eth_chainId') {
      ws.send(
        JSON.stringify({
          id: request.id,
          result: {
            result: CHAIN_ID,
          },
          jsonrpc: '2.0',
        })
      );
    }

    // Handle other methods
  });

  ws.on('error', console.error);

  ws.on('close', function () {
    // remove subscriptions from cache
  });
});
```

### Client side

```javascript
var subscription = web3.eth.subscribe(
  'logs',
  {
    address: '0x123456..',
    topics: ['0x12345...'],
  },
  function (error, result) {
    if (!error) console.log(result);
  }
);

> // unsubscribes the subscriptionsubscription.unsubscribe(function(error, success){

Reviewer: ```suggestion
// unsubscribes the subscription
subscription.unsubscribe(function(error, success){
```

if (success) {
  console.log('Successfully unsubscribed!');
}
```

### Mirror node polling:

1. Maintain a list of all live socket connections. This can be done with a `key-value` store where the `key` is a unique identifier (some ws libraries provide such an id, but if needed it can be generated by the relay) of the socket object, and the `value` is the socket object itself.

> 2. Maintain a list of all active subscriptions and the corresponding socket objects. It should have the following structure:

Reviewer: To confirm is this implying `count(live)` >= `count(active)`.
Also for every new subscription are we picking up from the list of remaining live connections. Meaning live connections always remain open and it's more of a queue system to assign and release usage of it?

Reviewer 3: A single socket connection can have 0, 1 or multiple subscriptions. It is created and destroyed when a client connects/disconnects from the ws-server. Subscriptions are created when `eth_subscribe` is called and are destroyed on `eth_unsubscribe` or when the associated connection is broken.

|     32609     |                                                 Invalid subscription id                                                  | Occurs when user tries to unsubscribe from a non-existing subscription or one that is not associated with the active socket connection. |
|     32610     |                                    Invalid subscription parameters: {detailed reason}                                    |               Occurs when there is an error due to the provided subscription parameters, i.e. the address does not exist.               |
|     32611     |                                                   Subscription aborted                                                   |                          Occurs when the subscription is aborted by the relay, i.e. a maximum TTL is reached.                           |

## Rate Limits

Active subscriptions should be limited in some way. We should create a flexible and configurable system for limiting subscriptions on a global and an IP level. This should be done by extending the existing `RateLimit` class as needed.

> 1. Add a configurable global limit on the total active subscriptions.

Reviewer: ```suggestion
1. Add a configurable global limit on the total active subscriptions. Initial default of 10.
```

> 2. Add a configurable limit on the total active subscriptions per IP.

Reviewer: ```suggestion
2. Add a configurable limit on the total active subscriptions per IP. Initial default of 1.
```

> 3. Add a TTL to subscriptions, i. e. automatically terminate a subscription after a configurable time period.

Reviewer: ```suggestion
3. Add a TTL to subscriptions, i. e. automatically terminate a subscription after a configurable time period. Initial default of 5 mins. 
```

## Tests

The following test cases should be covered but additional tests would be welcome.

1. Connecting to the ws-server works.
2. Subscribing to the `logs` event with all possible filter combinations - no filters, by address, by topics, multiple addresses, etc..
3. Unsubscribing from an active subscription.
4. Receiving the correct log data after an async process generates logs for that match/don't match the subscription filters.
5. Multiple active subscriptions and each of them receives their appropriate log data.
6. Exceeding the global subscription limit.
7. Exceeding the IP subscription limit.
8. Unsubscribing due to automatic subscription termination. 
9. E2E test using popular libraries (`ethers.js WebSocketProvider`) that include - connecting, subscribing, receiving data, unsubscribing

## Non-Functional Requirements

Users should be required to renew their subscription in the case of an error or if the subscription was closed from the relay side. If for some reason the relay is restarted it should not automatically try to restore all previous connections and subscriptions.

## Open Questions

> 1. How is the deployment going to be handled?

Reviewer: Do we have answers yet?
Initially we'll want the same container instance to serve these with a future option to split out as needed.
This means there should be an item to ensure container start manages both processes.
